state/couchbase/couchbase.go

```diff
@@ -0,0 +1,180 @@
+// ------------------------------------------------------------
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+// ------------------------------------------------------------
+
+package couchbase
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/dapr/components-contrib/state"
+	"gopkg.in/couchbase/gocb.v1"
+
+	jsoniter "github.com/json-iterator/go"
+)
+
+const (
+	couchbaseURL = "couchbaseURL"
+	username     = "username"
+	password     = "password"
+	bucket       = "bucket"
+)
+
+// Couchbase is a couchbase state store
+type Couchbase struct {
+	cbCluster *gocb.Cluster
+	bucket    string //TODO: having bucket sent as part of request (get,set etc.) metadata would be more flexible
+	json      jsoniter.API
+}
+
+// NewCouchbaseStateStore returns a new couchbase state store
+func NewCouchbaseStateStore() *Couchbase {
+	return &Couchbase{
+		json: jsoniter.ConfigFastest,
+	}
+}
+
+func validateMetadata(metadata state.Metadata) error {
+	if metadata.Properties[couchbaseURL] == "" {
+		return errors.New("couchbase error: couchbase URL is missing")
+	}
+
+	if metadata.Properties[username] == "" {
+		return errors.New("couchbase error: couchbase username is missing")
+	}
+
+	if metadata.Properties[password] == "" {
+		return errors.New("couchbase error: couchbase password is missing")
+	}
+
+	if metadata.Properties[bucket] == "" {
+		return errors.New("couchbase error: couchbase bucket is missing")
+	}
+
+	return nil
+}
+
+// Init does metadata and connection parsing
+func (cbs *Couchbase) Init(metadata state.Metadata) error {
+	err := validateMetadata(metadata)
+	if err != nil {
+		return err
+	}
+	cbs.bucket = metadata.Properties[bucket]
+	c, err := gocb.Connect(metadata.Properties[couchbaseURL])
+	if err != nil {
+		return fmt.Errorf("couchbase error: unable to connect to couchbase at %s - %v ", metadata.Properties[couchbaseURL], err)
+	}
+	cbs.cbCluster = c
+	//does not actually trigger the authentication
+	cbs.cbCluster.Authenticate(gocb.PasswordAuthenticator{
+		Username: metadata.Properties[username],
+		Password: metadata.Properties[password],
+	})
+	return nil
+}
+
+//Set stores value for a key to couchbase
+func (cbs *Couchbase) Set(req *state.SetRequest) error {
+	err := state.CheckSetRequestOptions(req)
+	if err != nil {
+		return err
+	}
+	var value string
+	b, ok := req.Value.([]byte)
+	if ok {
+		value = string(b)
+	} else {
+		value, err = cbs.json.MarshalToString(req.Value)
+	}
+
+	if err != nil {
+		return fmt.Errorf("couchbase error: failed to convert value %v", err)
+	}
+	//Under RBAC, bucket-passwords are no longer used
+	//see https://docs.couchbase.com/go-sdk/1.6/sdk-authentication-overview.html#authenticating-with-legacy-sdk-versions
+	bucket, err := cbs.cbCluster.OpenBucket(cbs.bucket, "")
```

Comment:
is `bucket` a cheap pass through object? If not, wouldn't it be better to create one on init? in the future when we enable a bucket to be specified per-call in the metadata, it could create an new bucket if needed.

Reply:
sure I think that can be done. bu that brings up another (generic) question - how does one close resources (connections etc.) spawned inside `Init()`?

Reply:
Dapr doesn't dispose of components unless the process is dead, so generally there's no external condition under which it would be needed to keep a reference to a component but close any of its internal resources.

Thinking about non-Dapr users, it might be useful in the future to add a `Close()` method for components.

Reply:
bucket is now created during `Init()`